### PR TITLE
Voxel size

### PR DESCRIPTION
**Pull request by jacg**
+ `voxelize_hits` was not respecting the requested `voxel_dimensions`.

+ There was no test of this behaviour.

This PR

+ adds a test demonstrating the problem
+ fixes `voxelize_hits` so that it passes the test
+ improves `__str__` in BHit and subclasses

This last point makes it much easer to understand and debug issues such as the main one in this PR.

### CODE REVIEW

**Comment by jacg** [comment]
The test proposed here is not correct: it fails to correctly express the idea "I want to make sure that this `float` is (approximately) a multiple of some other `float`".

I'm struggling to come up with a clean, concise, obviously correct (etc.) way of expressing this idea. On top of that, I keep getting the impression that I'm re-inventing some wheel.

Any suggestions how to do this well?

**Comment by gonzaponte** [comment]
> The test proposed here is not correct: it fails to correctly express the idea "I want to make sure that this float is (approximately) a multiple of some other float".

What is the `D` loop for? The `D` variable does nothing. If, as I feel, it is skippable, why do you say it does not express the idea correctly? It does *exactly* that.

**Comment by jacg** [comment]
Indeed, the `D`-loop is just leftover rubbish.

The non-rubbish does not express the idea correctly, because `5.9999999 % 1` is approximately `0.999999` which is nowhere near to `0`, even though `5.999999` is damn close to `6` which is a multiple of `1`.

In other words, if it is just *below* a multiple of whatever, then the test fails.

**Comment by gonzaponte** [comment]
I see. Why not:

```
np.allclose(distance_between_voxels, requested_voxel_dimensions)
```
?

**Comment by gonzaponte** [comment]
Sorry, didn't think too much about that. I see what you mean.

**Comment by gonzaponte** [comment]
The simplest ad-hoc way would be, I think:
```
discrete_distances = distance_between_voxels % requested_voxel_dimensions
np.allclose(distance_distances - np.round(discrete_distances), 0, rtol=eps)
```

for some `eps` much smaller than `requested_voxel_dimensions`. I guess the question is: Is there a simpler, more obvious way, right?

**Comment by paolafer** [comment]
Why don't you use the absolute value of the difference between the coordinates of two voxels minus voxel_size < epsilon?

**Comment by jacg** [comment]
Because this is guaranteed to fail for any pair of voxels that are not neighbours.

Is there any guarantee that successive elements of the voxel sequence returned by `voxelize_hits`, will be neighbours?

Even then, some neighbours will have identical centres in *some* dimensions, and then the scheme also fails.

**Comment by jacg** [comment]
I've pushed a failed attempt at fixing the test and the implementation. The test fails, so at least one of the two must be incorrect. I'll be almost completely unavailable for the next 40 hours. Please feel free to improve on my pathetic effort in the meantime.

**Comment by paolafer** [comment]
I can take this, Jacek!

**Comment by jjgomezcadenas** [comment]
You mean @jacg (aka Master) :)!

**Comment by paolafer** [comment]
I've tried to work on it, @jacg, and I found a solution that seems sensible to me and passes the tests (you can find it [here](https://github.com/paolafer/IC/blob/vox-size/invisible_cities/reco/paolina_functions.py)). The only problem is that it gives nonsense results, so there must be a flaw somewhere. I'll keep on working on this, I just wanted to push it in my branch, in case you want to have a look when you're back.

**Comment by jacg** [comment]
@paolafer In the version of the test that I see in your branch, the only `assert` in the test is commented out. In other words, the test will *always* pass. When I uncomment the loop with the assert, I get exactly the same failure as I see in my proposed solution.

**Comment by jacg** [comment]
There was a bug in the test I proposed earlier. I have fixed that now, and the proposed fixed implementation now passes the fixed test.

@paolafer Initial investigation suggests that the solution you propose in your branch does *not* pass the new test. So please check whether the solution proposed in **this** branch also gives 'nonsense results'.

I'm still not entirely happy with the test or the implementation: they seem too long-winded; but first let's check whether they make sense.

**Comment by paolafer** [comment]
@jacg I'm not sure I understand what you're telling me to do... In which branch did you do your changes? Your voxel-size branch seems to me the same as two days ago. What should I test with your branch?

**Comment by jacg** [comment]
> Your voxel-size branch seems to me the same as two days ago.

No, the branch has changed since then. Most importantly, `1` has been replaced with the voxel dimension in the second `isclose` in the test.

> What should I test with your branch?

Earier you said that 

> I found a solution that seems sensible to me and passes the tests (you can find it here). The only problem is that it gives nonsense results,

The solution you propose does not pass the test which is currently in `jacg/voxel-size`. Please fetch the latest version of `jacg/voxel-size` and use it to see whether it too gives you 'nonsense results' for whatever you meant when you said 'nonsense results'.

**Comment by paolafer** [comment]
Ok, now I understand.

**Comment by jacg** [comment]
@paolafer I've just pushed a change. See if you find this version of `voxelize_hits` easier to follow.

**Comment by paolafer** [comment]
Yes, this simplifies things!

**Comment by jacg** [comment]
As far as I am concerned, this one is now ready for review.

I would like the code to be a bit cleaner, but

+ it's already a bit better than the initial attempts
+ if it works, we should probably just take it and get on with other things
+ I suspect we might not be able to find a significantly crearer approach

**Comment by paolafer** [comment]
I would like to test it and compare the results with the previous version, @jerenner, maybe you want to try your code on this version, too.

**Comment by jacg** [comment]
> I would like to test it and compare the results with the previous version

Of course! I consider that to be a **necessary** part of the review.

**Comment by jacg** [comment]
You gotta love hypothesis!

In the `hard` profile hypothesis found the following counterexample for `test_voxelize_hits_does_not_lose_energy`:

    hits = [BHit(96.11764705882354, 10.0, 10.0, E=10.0), BHit(12.470588235294109, 10.100000000000016, 10.100000000000016, E=10.0)], voxel_dimensions = array([ 2.78823529,  1.        ,  1.        ])
    
        @given(bunch_of_hits, box_sizes)
        def test_voxelize_hits_does_not_lose_energy(hits, voxel_dimensions):
            voxels = voxelize_hits(hits, voxel_dimensions)
        
            if not hits:
                assert voxels == []
        
            def sum_energy(seq):
                return sum(e.E for e in seq)
        
    >       assert_almost_equal(sum_energy(hits), sum_energy(voxels))
    E       AssertionError: 
    E       Arrays are not almost equal to 7 decimals
    E        ACTUAL: 20.0
    E        DESIRED: 10.0


[Aside: with the way pytest reports *ACTUAL* and *DESIRED*, those arguments to `assert_almost_equal` should be swapped.]

So, what is going on? 

1. There are two hits, separated by an almost exact multiple of voxel lengths, in some direction.
2. The algorithm places the outer edges of the extreme voxels on the hit positions (as accurately as floating point operations allow).
3. One of the hits that sits on a voxel boundary is considered to lie outside the voxel, and it is lost, along with its energy!


Floating point numbers are such fun!

Any suggestion how to deal with this sensibly and painlessly?

**Comment by jacg** [comment]
Could you please help me to preserve my sanity? Try the following on your machine, show me the results and tell me if you think they make sense.

    from numpy import nextafter
    nextafter( 0.1, -1)
    nextafter( 0.1, +1)
    nextafter( 1  , -1)
    nextafter( 1  , +1)
    nextafter(10  , -1)
    nextafter(10  , +1)

**Comment by gonzaponte** [comment]
```
>>> from numpy import nextafter
>>> nextafter( 0.1, -1)
0.099999999999999992
>>> nextafter( 0.1, +1)
0.10000000000000002
>>> nextafter( 1  , -1)
0.99999999999999989
>>> nextafter( 1  , +1)
1.0
>>> nextafter(10  , -1)
9.9999999999999982
>>> nextafter(10  , +1)
9.9999999999999982
```
Ha! It is not only assymetric, but inconsistent!

**Comment by jacg** [comment]
Bug or feature?

**Comment by gonzaponte** [comment]
They do have a [test](https://github.com/numpy/numpy/blob/master/numpy/core/tests/test_half.py#L249) about it, but I don't quite understand it. The array they create doesn't make sense to me.

**Comment by jacg** [comment]
OK, here's the relevant part of the docs: 

> Return the next floating-point value after x1 towards x2, element-wise.

the key part being **towards x2**. (Unfortunately this is written right at the top of the docs, so looking for the meaning of `x2` in the documentation for the `x2` parameter didn't give me any joy.)

So, instead of `+1` it requires something like `+np.inf`

**Comment by jacg** [comment]
Well,

    eps = 1e-12
    voxel_edges_lo -= eps
    voxel_edges_hi += eps

seems to do the trick, for values of `eps` ranging from about `1e-9` to `1e-14`.

But it's not a solution that I'm particularly happy with. It seems like a hack.

I should really try to understand the exact circumstances under which the problem arises, but my time available for IC is going to be very limited until the end of the week, and essentially zero for the week-and-a-half after that. So I'm probably dropping off the radar for a while.

**Comment by jacg** [comment]
I have pushed the aforementioned solution.

I weep for `voxelize_hits` who was fairly slim in its youth, but is now becoming rather obese.

**Comment by paolafer** [comment]
Just to confirm, the current version or your voxel-size branch should have this problem fixed, right, @jacg? I will proceed to test it.

**Comment by jacg** [comment]
Yes, the fix is there. How general the fix is, I do not know, as I haven't analysed the problem to fully understand it.

Also, it shouldn't make much difference in real life, because the frequency with which this problem arises should be absolutely miniscule: how likely is it that two extreme voxels are separated by 'exactly' a multiple of the voxel width?

**Comment by paolafer** [comment]
Yes, it should probably make no difference.

**Comment by jjgomezcadenas** [comment]
@paolafer do not let this PR to grow a beard

**Comment by paolafer** [comment]
I'm on it!

**Review by paolafer — COMMENTED** [review]
Could you please clarify to me what the assertion of test_voxelize_hits_respect_voxel_dimensions means? I understand that the remainder of the division should be 0, but why unit?

**Comment by jacg** [comment]
`unit` is just a short synonym for `requested_voxel_dimensions`. As discussed above in this thread, we want the dimensions of the voxels that were returned to be (very close to) a multiple of this unit size. The test does this by looking at the remainder when dividing by the unit size.

The problem is that both `1.000000001` times the unit size and `5.999999999` times the unit size **should** succeed, but the latter evaluates to `0.9999999999` and will fail if you only admit `off_by % 0`: you need `off_by % unit` for that one.

In summary, the second case is needed for those values which are approximately close from below, because the first case only spots those which are approximately close from above.

**Review by paolafer — APPROVED** [review]
This PR fixes a bug in the dimension of the voxels. Now voxels have all the same dimensions. In my opinion it can be approved and merged.

**Comment by jacg** [comment]
The voxels all had the same dimensions before ... it's just that those dimensions were usually slightly smaller than the requested ones.

The code in this PR leaves a margin in the outer voxels (in order to match the requested voxel size exactly). Consequently the blobs will have slightly lower energy than before.

**Comment by paolafer** [comment]
As far as I recall, the first changes in this PR have been done to address the fact that the positions of the voxels were not separated by a multiple of the voxel dimension. Maybe I'm confused and that problem was solved by another PR?

**Comment by jacg** [comment]
Yes, this PR addresses the fact that the returned voxels were not of the specified required dimension, but slightly smaller. You earlier statement (emphasis mine)

> **Now** voxels have all the same dimensions.

is misleading, because it suggests that before this PR the voxels did *not* have the same dimensions. This is not true: they *did* have the same dimensions *as eachother* but those dimensions were slightly smaller than the requested dimensions.

**Comment by paolafer** [comment]
Ok, it's all clear! I have used the code and it produces voxels with the requested dimensions. Green light from my side!
